source/slang/slang-legalize-types.cpp

```diff
@@ -203,6 +203,25 @@ bool isResourceType(IRType* type)
     return false;
 }
 
+// Helper wrapper function around isResourceType that checks not only if the
+// given type is a resourceType, but if it is also a pointer to a resourceType.
+bool hasResourceType(IRType* type)
+{
+    if (isResourceType(type))
+        return true;
+    while (auto ptrType = as<IRPtrTypeBase>(type))
+    {
+        type = ptrType->getValueType();
+    }
+
+    // Special handling for RWStructuredBuf
+    if (const auto rwStructBufferType = as<IRStructType>(type))
```

Comment:
Struct type is not rwstructured buffer. You should be looking for IRHLSLStructuredBufferTypeBase.

Reply:
The trace from debugger shows that RWStructBuf is saved as a `kIROp_StructType`
```
| type | 0x000001b38ef58150 {kIROp_StructType (102) 0x8ef58150} | Slang::IRType *
   | [op] | kIROp_StructType (102) | Slang::IROp
   | [UID] | 1093 | unsigned int
     | [type] | 0x000001b38ef580e0 {kIROp_TypeKind (48) 0x8ef580e0} | Slang::IRInst *
     | [name] | "RWStructuredBuffer" | char[18]
```

Reply:
It is eventually lowered into a struct during legalize spirv or some legalize buffer pass. We need to make sure this is handled before they are lowered into structs to avoid doing incorrect handling with other ordinary structs.

Reply:
It is converted from `IRHLSLStructuredBufferTypeBase` into `IRStructType` in `processGlobalParam:lowerStructuredBufferType` within `legalizeSPIRV:processModue`
right before `processWorkList` where we deal with the `nonUniformResourceInfo`.

Is it ok to not do the struct conversion, if it's user (GEP) performs a NonUniformResource access? That way, we still have a rwStructBuffer type

For clarity, this is the sequence I'm talking about:
```
        {502}   let  %24        : UInt  = nonUniformResourceIndex(%23)
        [{541}  nameHint("buffer1")]
        {538}   let  %buffer1_  : RWStructuredBuffer(Int, DefaultLayout, %13)   = getElement(%buffer, %24)
        {544}   let  %25        : Ptr(Int)      = rwstructuredBufferGetElementPtr(%buffer1_, 0 : UInt)
```

After `lowerStructuredBufferType`, this gets converted into:
```
        {502}   let  %26        : UInt  = nonUniformResourceIndex(%25)
        {1002}  let  %27        : Ptr(%RWStructuredBuffer, 12 : Int)    = getElementPtr(%buffer, %26)
        {544}   let  %28        : Ptr(Int)      = rwstructuredBufferGetElementPtr(%27, 0 : UInt)
}
```

Reply:
Type legalization happens way before spirv legalization so you should never need to worry about this in slang-legalize-types.cpp. When in this file, you can assume all structured buffer parameters are having the IRHLSLStructuredBufferBase type.

Reply:
Ah I see your point. Even though the file is type-legalize, it is actually hijacked and being called from spirv-legalize. That is why, by that time, the original type is lost.

Reply:
by making a more explicit check for struct buf, like `ptrType->getAddressSpace() == SpvStorageClassStorageBuffer` we can move past this problem.